src/Aspire.Hosting/ApplicationModel/InteractionService.cs

```diff
@@ -193,7 +193,41 @@ internal async Task CompleteInteractionAsync(int interactionId, Func<Interaction
             }
         }
 
-        var result = await createResult(interactionState, _serviceProvider, cancellationToken).ConfigureAwait(false);
+        var result = createResult(interactionState, _serviceProvider, cancellationToken);
+
+        // Run validation for inputs interaction.
+        if (result.Complete && interactionState.InteractionInfo is Interaction.InputsInteractionInfo inputsInfo)
+        {
+            if (result.State is IReadOnlyList<InteractionInput> inputs)
+            {
+                var options = (InputsDialogInteractionOptions)interactionState.Options;
+
+                if (options.ValidationCallback is { } validationCallback)
+                {
+                    foreach (var input in inputs)
+                    {
+                        input.ValidationErrors.Clear();
+                    }
+
+                    var context = new InputsDialogValidationContext
+                    {
+                        CancellationToken = cancellationToken,
+                        ServiceProvider = _serviceProvider,
+                        Inputs = inputsInfo.Inputs
+                    };
+                    await validationCallback(context).ConfigureAwait(false);
+
+                    if (context.HasErrors)
+                    {
+                        result = new InteractionCompletionState { Complete = false, State = inputs };
+                    }
+                }
+            }
+            else
+            {
+                throw new InvalidOperationException($"Expected state of {typeof(IReadOnlyList<InteractionInput>).FullName} for completed inputs interaction.");
```

Comment:
Should this throw? Who will see it?

Reply:
It should throw. I expect the error would bubble up in the app host.

Reply:
It already is: https://github.com/dotnet/aspire/blob/2cc7fb1c2057dc686fee781770cb73c62e2153b2/src/Aspire.Hosting/Publishing/PublishingActivityProgressReporter.cs#L332-L336